server/datastore/s3/common_file_store.go

```diff
@@ -130,13 +131,36 @@ func (s *commonFileStore) Cleanup(ctx context.Context, usedFileIDs []string, rem
 		return 0, nil
 	}
 
-	res, err := s.s3client.DeleteObjects(&s3.DeleteObjectsInput{
-		Bucket: &s.bucket,
-		Delete: &s3.Delete{
-			Objects: toDeleteKeys,
-		},
-	})
-	return len(res.Deleted), ctxerr.Wrapf(ctx, err, "deleting %s in S3 store", s.fileLabel)
+	deleted := 0
+	var errs []error
+	var g errgroup.Group
+	g.SetLimit(10)
+
+	for _, obj := range toDeleteKeys {
+		obj := obj
+		g.Go(func() error {
+			_, err := s.s3client.DeleteObject(&s3.DeleteObjectInput{
+				Bucket: &s.bucket,
+				Key:    obj.Key,
+			})
+			if err != nil {
+				errs = append(errs, ctxerr.Wrapf(ctx, err, "deleting %s in S3 store", s.fileLabel))
```

Comment:
Is this concurrency safe in a go-routine?

Reply:
@ksykulev I think `errgroup` aims to return just the first error that it gets back from the goroutines, so instead of appending errors to a slice, you can just return the error and then check the return value from `g.Wait` like you do below.